hspec-discover/test/Test/Hspec/Discover/SortSpec.hs

```diff
@@ -0,0 +1,49 @@
+module Test.Hspec.Discover.SortSpec (main, spec) where
+
+import           Helper
+
+import           Data.List (sortBy)
+import           Data.Ord (comparing)
+
+import           Test.Hspec.Discover.Sort
+
+main :: IO ()
+main = hspec spec
+
+
+spec :: Spec
+spec = do
+  describe "naturalSortKey" $ do
+    context "compares" $ do
+      it "reflexively" $ do
+        "" <=> "" `shouldBe` EQ
+        "Hello2World" <=> "Hello2World" `shouldBe` EQ
+        "123Hello456" <=> "123Hello456" `shouldBe` EQ
+        "" <=> "" `shouldBe` EQ
+      it "empty first" $ do
+        "" <=> "Hello2World" `shouldBe` LT
+        "Hello2World" <=> "" `shouldBe` GT
+      it "numbers first" $ do
+        "Hello2World" <=> "Hello World" `shouldBe` LT
+        "Hello World" <=> "Hello2World" `shouldBe` GT
+      it "string segments" $ do
+        "Hello2World9" <=> "Hello2World!0" `shouldBe` LT
+        "Hello2World!0" <=> "Hello2World9" `shouldBe` GT
+      it "numeric segments" $ do
+        "3.1.415" <=> "3.14.15" `shouldBe` LT
+        "3.14.15" <=> "3.1.415" `shouldBe` GT
```

Comment:
You may be able to make some of the things you test here redundant by deriving the `Ord` instance, e.g.:

```haskell
import           Data.Char

data NaturalSortChunk = NumericChunk Integer Int | StringChunk String
  deriving (Eq, Show, Ord)

naturalSortKey :: String -> [NaturalSortChunk]
naturalSortKey s@(c : _)
  | isDigit c = NumericChunk (read num) (length num) : naturalSortKey afterNum
  | otherwise = StringChunk str : naturalSortKey afterStr
  where
    (num, afterNum) = span  isDigit s
    (str, afterStr) = break isDigit  s
naturalSortKey [] = []
```

Reply:
Ah sure, I suppose since `isDigit` is ASCII digits only it's not like there can be ties that need to be broken by string.

Reply:
The code I give still uses the string length as a tie breaker.  This is why is `NumericChunk Integer Int`.  The `Int` is the length of the string.

Reply:
And yes, there are ties, e.g. `01` and `1`, see below.

Reply:
I meant, broken by string comparison in addition to by length. That could happen with Unicode digits but (numeric value, length) should be sufficient with ASCII only.